rust/src/core.rs

```diff
@@ -77,15 +83,35 @@ pub(crate) struct FilesystemScriptPrep {
 
 pub(crate) fn prepare_filesystem_script_prep(rootfs: i32) -> CxxResult<Box<FilesystemScriptPrep>> {
     let rootfs = ffi_view_openat_dir(rootfs);
-    rootfs.local_rename_optional(SSS_CACHE, &format!("{}.orig", SSS_CACHE))?;
-    Ok(Box::new(FilesystemScriptPrep { rootfs }))
+    Ok(FilesystemScriptPrep::new(rootfs)?)
 }
 
 impl FilesystemScriptPrep {
+    /// Filesystem paths that we rename out of the way if present
+    const OPTIONAL_PATHS: &'static [&'static str] = &[SSS_CACHE_PATH, SYSTEMCTL_PATH];
+
+    fn saved_name(name: &str) -> String {
+        format!("{}.rpmostreesave", name)
+    }
+
+    pub(crate) fn new(rootfs: openat::Dir) -> Result<Box<Self>> {
+        for &path in Self::OPTIONAL_PATHS {
+            rootfs.local_rename_optional(path, &Self::saved_name(path))?;
+        }
+        // And actually replace the original systemctl with our wrapper, if we found
+        // our saved version.
+        if rootfs.exists(&Self::saved_name(SYSTEMCTL_PATH))? {
+            rootfs.write_file_contents(SYSTEMCTL_PATH, 0o755, SYSTEMCTL_WRAPPER)?;
+        }
```

Comment:
I'm confused by this. Shouldn't it be something like:

```rust
if rootfs.local_rename_optional(SYSTEMCTL_PATH, &Self::saved_name(SYSTEMCTL_PATH))? {
    rootfs.write_file_contents(SYSTEMCTL_PATH, 0o755, SYSTEMCTL_WRAPPER)?;
    
}
```

and then in the `undo` path we rename it back?

Reply:
I guess that'll require also adding a e.g. `renamed: Vec<String>` member to `FilesystemScriptPrep` for the `undo` path.

Reply:
Yeah I was trying to avoid special casing `SYSTEMCTL_PATH`.  I believe the logic is correct, but I agree it's not obvious.  The flow is:

- `rename("systemctl", "systemctl.rpmostreesave")`
- `stat("systemctl.rpmostreesave")` == exists
- `write(systemctl, "new contents")`

Then:

`rename("systemctl.rpmostreesave", "systemctl")`

Reply:
> I guess that'll require also adding a e.g. renamed: Vec<String> member to FilesystemScriptPrep for the undo path.

Right.  One way to think about this is we're effectively serializing the "file existed" boolean into the filesystem itself.